File: example/서로소 집합/크루스칼/크루스칼.py
# 신장 트리는 하나의 그래프가 모든 노드를 포함하면서 사이클이 존재하지 않는 그래프
# 최소 신장 트리 알고리즘은 모든 노드를 연결하면서 최소 비용인 경우를 찾는 알고리즘
# 크루스칼 알고리즘은 가장 거리가 짧은 간선부터 집합에 포함시키고 사이클이 돌지 않게 함

import sys
input = sys.stdin.readline


def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]


def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b


v, e = map(int, input().split())
parent = [0]*(v+1)
edges = []
result = 0

for i in range(1, v+1):
    parent[i] = i

for _ in range(e):
    a, b, cost = map(int, input().split())
    edges.append((cost, a, b))


# 가장 작은 비용부터 선택
edges.sort()

for edge in edges:
    cost, a, b = edge
    if find_parent(parent, a) != find_parent(parent, b):
        union_parent(parent, a, b)
        result += cost

print(result)

File: java/src/algorithm/서로소_집합/최소_신장_트리/크루스칼.java
package algorithm.서로소_집합.최소_신장_트리;

import java.util.ArrayList;
import java.util.List;

// 최소 신장 트리(Minimum Spanning Tree, MST): 그래프의 모든 노드를 포함하면서 사이클이 존재하지 않는 부분 그래프
// 크루스칼 알고리즘: 가장 적은 비용으로 신장 트리를 찾는 알고리즘
public class 크루스칼 {

    static final int V = 7;
    static final int E = 9;
    static final int[][] INPUT = {
            {1, 2, 29},
            {1, 5, 75},
            {2, 3, 35},
            {2, 6, 34},
            {3, 4, 7},
            {4, 6, 23},
            {4, 7, 13},
            {5, 6, 53},
            {6, 7, 25},
    };

    static int[] parent = new int[V + 1];
    static List<Edge> edges = new ArrayList<>();
    static int result = 0;

    public static void main(String[] args) {
        // 부모 테이블을 자기 자신으로 초기화
        for (int i = 1; i <= V; i++) {
            parent[i] = i;
        }

        // 간선 정보 입력 받기
        for (int i = 0; i < E; i++) {
            int from = INPUT[i][0];
            int to = INPUT[i][1];
            int weight = INPUT[i][2];

            edges.add(new Edge(from, to, weight));
        }

        // 간선의 비용을 기준으로 오름차순 정렬
        edges.sort((a, b) -> a.weight - b.weight);

        for (Edge edge : edges) {
            int from = edge.from;
            int to = edge.to;
            int weight = edge.weight;

            // 사이클이 발생하지 않는 경우에만 집합에 포함
            // 간선의 노드에 대해서 union 하기 전에 parent 가 같다면 사이클이 발생한다는 의미
            if (findParent(from) != findParent(to)) {
                unionParent(from, to);
                result += weight;
            }
        }

        System.out.println(result);
    }

    public static int findParent(int x) {
        if (x == parent[x]) {
            return x;
        }

        return parent[x] = findParent(parent[x]);
    }

    public static void unionParent(int a, int b) {
        a = findParent(a);
        b = findParent(b);

        if (a < b) {
            parent[b] = a;
        } else {
            parent[a] = b;
        }
    }
}

class Edge {

    int from;
    int to;
    int weight;

    public Edge(int from, int to, int weight) {
        this.from = from;
        this.to = to;
        this.weight = weight;
    }
}